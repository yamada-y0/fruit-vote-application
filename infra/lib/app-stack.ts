import * as cdk from "aws-cdk-lib";
import { StackProps } from "aws-cdk-lib";
import * as ecr from "aws-cdk-lib/aws-ecr";
import { Construct } from "constructs";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as ecs from "aws-cdk-lib/aws-ecs";
import * as ecs_patterns from "aws-cdk-lib/aws-ecs-patterns";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import { BillingMode } from "aws-cdk-lib/aws-dynamodb";

export interface AppStackProps extends StackProps {
    appName: string;
    vpc: ec2.IVpc;
}

export class AppStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props: AppStackProps) {
        super(scope, id, props);
        const repository = new ecr.Repository(this, "EcrRepository", {
            repositoryName: `${props.appName}-app`,
            removalPolicy: cdk.RemovalPolicy.DESTROY,
            emptyOnDelete: true,
        });

        const cluster = new ecs.Cluster(this, "EcsCluster", {
            vpc: props.vpc,
            clusterName: `${props.appName}-cluster`,
        });
        const fargateService =
            new ecs_patterns.ApplicationLoadBalancedFargateService(
                this,
                "FargateService",
                {
                    cluster: cluster,
                    cpu: 256,
                    desiredCount: 1,
                    taskImageOptions: {
                        image: ecs.ContainerImage.fromEcrRepository(repository),
                        containerPort: 8080,
                        enableLogging: true,
                    },
                    memoryLimitMiB: 512,
                    publicLoadBalancer: false,
                },
            );

        const voteTable = new dynamodb.Table(this, "VoteTable", {
            partitionKey: {
                name: "userId",
                type: dynamodb.AttributeType.STRING,
            },
            sortKey: {
                name: "fruit",
                type: dynamodb.AttributeType.STRING,
            },
            tableName: "Vote",
            removalPolicy: cdk.RemovalPolicy.DESTROY,
            billingMode: BillingMode.PAY_PER_REQUEST,
        });
        voteTable.grantReadWriteData(
            fargateService.service.taskDefinition.taskRole,
        );

        const fruitCountTable = new dynamodb.Table(this, "FruitCountTable", {
            partitionKey: {
                name: "fruit",
                type: dynamodb.AttributeType.STRING,
            },
            tableName: "FruitCount",
            removalPolicy: cdk.RemovalPolicy.DESTROY,
            billingMode: BillingMode.PAY_PER_REQUEST,
        });
        fruitCountTable.grantReadWriteData(
            fargateService.service.taskDefinition.taskRole,
        );
    }
}
