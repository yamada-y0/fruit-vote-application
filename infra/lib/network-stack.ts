import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as ec2 from "aws-cdk-lib/aws-ec2";

export class NetworkStack extends cdk.Stack {
    public readonly vpc: ec2.IVpc;

    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);
        this.vpc = new ec2.Vpc(this, "VPC", {
            ipAddresses: ec2.IpAddresses.cidr("10.0.0.0/16"),
            subnetConfiguration: [
                {
                    cidrMask: 24,
                    name: "Isolated",
                    subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
                },
            ],
            maxAzs: 2,
        });

        new ec2.InterfaceVpcEndpoint(this, "EcrApiEndpoint", {
            vpc: this.vpc,
            service: ec2.InterfaceVpcEndpointAwsService.ECR,
        });
        new ec2.InterfaceVpcEndpoint(this, "EcrDockerEndpoint", {
            vpc: this.vpc,
            service: ec2.InterfaceVpcEndpointAwsService.ECR_DOCKER,
        });
        new ec2.InterfaceVpcEndpoint(this, "CloudWatchLogsEndpoint", {
            vpc: this.vpc,
            service: ec2.InterfaceVpcEndpointAwsService.CLOUDWATCH_LOGS,
        });
        new ec2.GatewayVpcEndpoint(this, "S3Endpoint", {
            vpc: this.vpc,
            service: ec2.GatewayVpcEndpointAwsService.S3,
        });
        new ec2.GatewayVpcEndpoint(this, "DynamoDbEndpoint", {
            vpc: this.vpc,
            service: ec2.GatewayVpcEndpointAwsService.DYNAMODB,
        });
    }
}
