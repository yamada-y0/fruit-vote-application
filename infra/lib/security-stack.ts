import * as cdk from 'aws-cdk-lib';
import {Duration} from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import {Construct} from 'constructs';

export class SecurityStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);
        const provider = new iam.OpenIdConnectProvider(this, 'GitHubOIDC', {
            url: 'https://token.actions.githubusercontent.com',
            clientIds: ['sts.amazonaws.com'],
        })
        const role = new iam.Role(this, 'GitHubActionsRole', {
            roleName: 'GitHubActionsRole',
            maxSessionDuration: Duration.hours(2),
            assumedBy: new iam.WebIdentityPrincipal(provider.openIdConnectProviderArn, {
                StringEquals: {
                    'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
                },
                StringLike: {
                    'token.actions.githubusercontent.com:sub': `repo:yamada-y0/fruit-vote-benchmark:*`,
                }
            }),
        })
        role.addManagedPolicy(
            iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonEC2ContainerRegistryPowerUser'),
        )
    }
}
