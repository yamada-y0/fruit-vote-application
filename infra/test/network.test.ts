import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as Network from "../lib/network-stack";

describe("NetworkStack Test", () => {
    test("Snapshot Test", () => {
        const app = new cdk.App();
        // WHEN
        const stack = new Network.NetworkStack(app, "NetworkTestStack");

        // THEN
        const template = Template.fromStack(stack);
        expect(template.toJSON()).toMatchSnapshot();
    });
});
