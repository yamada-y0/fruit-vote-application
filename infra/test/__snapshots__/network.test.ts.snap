// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkStack Test Snapshot Test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CloudWatchLogsEndpointF23EDC4F": {
      "Properties": {
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "CloudWatchLogsEndpointSecurityGroup3E72E04B",
              "GroupId",
            ],
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".logs",
            ],
          ],
        },
        "SubnetIds": [
          {
            "Ref": "VPCIsolatedSubnet1SubnetEBD00FC6",
          },
          {
            "Ref": "VPCIsolatedSubnet2Subnet4B1C8CAA",
          },
        ],
        "VpcEndpointType": "Interface",
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "CloudWatchLogsEndpointSecurityGroup3E72E04B": {
      "Properties": {
        "GroupDescription": "NetworkTestStack/CloudWatchLogsEndpoint/SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::GetAtt": [
                "VPCB9E5F0B4",
                "CidrBlock",
              ],
            },
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "VPCB9E5F0B4",
                      "CidrBlock",
                    ],
                  },
                  ":443",
                ],
              ],
            },
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "DynamoDbEndpointBF7E74F2": {
      "Properties": {
        "RouteTableIds": [
          {
            "Ref": "VPCIsolatedSubnet1RouteTableEB156210",
          },
          {
            "Ref": "VPCIsolatedSubnet2RouteTable9B4F78DC",
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".dynamodb",
            ],
          ],
        },
        "VpcEndpointType": "Gateway",
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "EcrApiEndpointB01DFFD7": {
      "Properties": {
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "EcrApiEndpointSecurityGroupE2E875BD",
              "GroupId",
            ],
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".ecr.api",
            ],
          ],
        },
        "SubnetIds": [
          {
            "Ref": "VPCIsolatedSubnet1SubnetEBD00FC6",
          },
          {
            "Ref": "VPCIsolatedSubnet2Subnet4B1C8CAA",
          },
        ],
        "VpcEndpointType": "Interface",
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "EcrApiEndpointSecurityGroupE2E875BD": {
      "Properties": {
        "GroupDescription": "NetworkTestStack/EcrApiEndpoint/SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::GetAtt": [
                "VPCB9E5F0B4",
                "CidrBlock",
              ],
            },
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "VPCB9E5F0B4",
                      "CidrBlock",
                    ],
                  },
                  ":443",
                ],
              ],
            },
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "EcrDockerEndpointE4437377": {
      "Properties": {
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "EcrDockerEndpointSecurityGroup5A4BC357",
              "GroupId",
            ],
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".ecr.dkr",
            ],
          ],
        },
        "SubnetIds": [
          {
            "Ref": "VPCIsolatedSubnet1SubnetEBD00FC6",
          },
          {
            "Ref": "VPCIsolatedSubnet2Subnet4B1C8CAA",
          },
        ],
        "VpcEndpointType": "Interface",
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "EcrDockerEndpointSecurityGroup5A4BC357": {
      "Properties": {
        "GroupDescription": "NetworkTestStack/EcrDockerEndpoint/SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::GetAtt": [
                "VPCB9E5F0B4",
                "CidrBlock",
              ],
            },
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "VPCB9E5F0B4",
                      "CidrBlock",
                    ],
                  },
                  ":443",
                ],
              ],
            },
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "S3EndpointD570F362": {
      "Properties": {
        "RouteTableIds": [
          {
            "Ref": "VPCIsolatedSubnet1RouteTableEB156210",
          },
          {
            "Ref": "VPCIsolatedSubnet2RouteTable9B4F78DC",
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".s3",
            ],
          ],
        },
        "VpcEndpointType": "Gateway",
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VPCB9E5F0B4": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "NetworkTestStack/VPC",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VPCIsolatedSubnet1RouteTableAssociationA2D18F7C": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VPCIsolatedSubnet1RouteTableEB156210",
        },
        "SubnetId": {
          "Ref": "VPCIsolatedSubnet1SubnetEBD00FC6",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCIsolatedSubnet1RouteTableEB156210": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "NetworkTestStack/VPC/IsolatedSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCIsolatedSubnet1SubnetEBD00FC6": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          {
            "Key": "Name",
            "Value": "NetworkTestStack/VPC/IsolatedSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCIsolatedSubnet2RouteTable9B4F78DC": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "NetworkTestStack/VPC/IsolatedSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCIsolatedSubnet2RouteTableAssociation7BF8E0EB": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VPCIsolatedSubnet2RouteTable9B4F78DC",
        },
        "SubnetId": {
          "Ref": "VPCIsolatedSubnet2Subnet4B1C8CAA",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCIsolatedSubnet2Subnet4B1C8CAA": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.1.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          {
            "Key": "Name",
            "Value": "NetworkTestStack/VPC/IsolatedSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
